Fixed getting PE ratios from MSN

losgrandes opened the pull request:
I've changed the way PE ratios are fetched from MSN Money based on the findings in https://github.com/mrhappyasthma/IsThisStockGood/issues/57.

losgrandes committed:
Fixed getting PE ratios from MSN

losgrandes commented on src/DataFetcher.py:
```diff
@@ -165,20 +165,35 @@ def parse_stockrow_key_stats(self, response, *args, **kwargs):
   def parse_pe_ratios(self, response, *args, **kwargs):
     if response.status_code != 200:
       return
-    if not self.pe_ratios:
```
I don't think this is possible for `self.pe_ratios` to not exist. `self.pe_ratios = MSNMoney(self.ticker_symbol)` will always make this assumption invalid.

mrhappyasthma replied:
We clear out the self.pe_ratios on failure of this method. (line 198 on the right-hand diff).

So it's probably best to keep this check, on the off chance someone attempts to call in to this. Just to avoid a crash.

losgrandes replied:
Reverted

mrhappyasthma commented on src/MSNMoney.py:
```diff
@@ -1,46 +1,38 @@
-    url = MSNMoney.BASE_URL.format(ticker_symbol)
-    return url
+  TICKER_URL = 'https://services.bingapis.com/contentservices-finance.csautosuggest/api/v1/Query?query={}&market=en-us'
+  BASE_URL = 'https://services.bingapis.com/contentservices-finance.financedataservice/api/v1/KeyRatios?stockId={}'
```
This isn't really a base URL anymore. Can we instead call this KEY_RATIOS_URL or something?

losgrandes replied:
Done

mrhappyasthma commented on src/MSNMoney.py:
```diff
@@ -1,46 +1,38 @@
+
+  def extract_stock_id(self, content):
+    data = json.loads(content)
+    for ticker in data['data']['stocks']:
```
To make these more fault-tolerant against URL change, can we make them in to `.get()` instead of using the `[]` syntax?

We can return a default empty dict in those cases.

`data.get('data', {}).get('stocks', [])`

--

Same with the other keys `RT00S` and `SecId`.

losgrandes replied:
Improved

mrhappyasthma commented on src/MSNMoney.py:
```diff
@@ -1,46 +1,38 @@
+      for year in data['companyMetrics']
+      if year['fiscalPeriodType'] == 'Annual'
+      and 'priceToEarningsRatio' in year
+    ][-5:]
```
Can we turn this -5 in to a constant variable so it's a bit more readable and easier to change (if we ever want to find the min/max over a larger time span).

e.g.

pe_ratio_years_span (or better name)?

losgrandes replied:
Done

mrhappyasthma commented on src/MSNMoney.py:
```diff
@@ -1,46 +1,38 @@
+      year['priceToEarningsRatio']
+      for year in data['companyMetrics']
+      if year['fiscalPeriodType'] == 'Annual'
+      and 'priceToEarningsRatio' in year
```
Let's use the `.get(...., <default>)` here as well. Just to avoid issues if the data or data format ever changes.

losgrandes replied:
Done

mrhappyasthma commented:
Generally the code looks good. Thanks again for doing this.

I have to patch it locally to test it still, but looks sound to me. Left some comments, but mostly small stuff.

mrhappyasthma commented on src/MSNMoney.py:
```diff
@@ -1,46 +1,38 @@
+    data = json.loads(content)
+    for ticker in data['data']['stocks']:
+        js = json.loads(ticker)
+        if js['RT00S'] == self.ticker_symbol:
```
I found one bug: this needs to be `self.ticker_symbol.upper()`, as the values in `RT005` are always the capitalized versions. otherwise `AMZN` will not match `amzn`.

losgrandes replied:
Done

mrhappyasthma commented:
Aside from the case sensitivity issue, this appears to work as well as I'd hoped when testing locally :)

losgrandes committed:
Improved namings in MSNMoney
Fault-tolerant PE Ratios extraction
Added tests for MSNMoney's _parse_pe_ratios

losgrandes committed:
Don't calculate prices if less that 5 years of PE ratios

losgrandes commented:
@mrhappyasthma All your suggestions have been applied. I've added unit tests for `_parse_pe_ratios`.

However, something is off, cause I get very different results for e.g. GOOG within several minutes. Seems like something random.

Now I'm geting for GOOG:
```
margin_of_safety_price: 26.6402
         sticker_price: 53.2804
         current_price: 97.6
```

Several minutes ago:
```
margin_of_safety_price: 391.2
         sticker_price: 782.4
         current_price: 97.6
```

I don't think there's anything wrong with PE ratios. I'll try to investigate this. Without knowing why this PR cannot be merged.

losgrandes commented:
> I don't think there's anything wrong with PE ratios. I'll try to investigate this. Without knowing why this PR cannot be merged.

I reproduced this, but only when using Rule#1 Stock Screener. Let me look at what's off there.

mrhappyasthma commented:
Could you clarify the repro steps for the issue you noticed with GOOG?

Were you just using the site and querying GOOG and seeing different results at different times?

(I'd be a bit surprised as I imagine most of the data should be pretty static. But perhaps there's either a bug somewhere or one of the data sources that we scrape is unreliable.)

mrhappyasthma commented:
That said, I think the code here is sound. So I'm gonna merge it in. And if we can observe a reliable way to reproduce, we can debug further.

losgrandes commented:
> Could you clarify the repro steps for the issue you noticed with GOOG?
> 
> Were you just using the site and querying GOOG and seeing different results at different times?
> 
> (I'd be a bit surprised as I imagine most of the data should be pretty static. But perhaps there's either a bug somewhere or one of the data sources that we scrape is unreliable.)

https://github.com/mrhappyasthma/IsThisStockGood/issues/50#issuecomment-1328208406
